src/dbnode/storage/index.go

```diff
@@ -2066,6 +2066,114 @@ func (i *nsIndex) CleanupExpiredFileSets(t time.Time) error {
+		ReaderBufferSize: fsOpts.InfoReaderBufferSize(),
+		IncludeCorrupted: true,
+	})
+
+	if len(infoFiles) == 0 {
+		return nil
+	}
+
+	var (
+		toDelete []string
+		begin    = 0 // marks the beginning of a subslice that contains filesets with same block starts
+	)
+	// It's expected that info files are ordered by block start and volume index
+	for j := range infoFiles {
+		if infoFiles[begin].ID.BlockStart.Before(infoFiles[j].ID.BlockStart) {
+			files, err := i.getCorruptedVolumesForDeletion(infoFiles[begin:j])
+			if err != nil {
+				return err
+			}
+			toDelete = append(toDelete, files...)
+			begin = j
+		} else if infoFiles[begin].ID.BlockStart.After(infoFiles[j].ID.BlockStart) {
+			errorMessage := "filesets are expected to be ordered by block start"
+			instrument.EmitAndLogInvariantViolation(i.opts.InstrumentOptions(), func(l *zap.Logger) {
+				l.Error(errorMessage)
+			})
+			return instrument.InvariantErrorf(errorMessage)
+		}
+	}
+
+	// Process the volumes in the last block, which are not covered by the loop.
+	files, err := i.getCorruptedVolumesForDeletion(infoFiles[begin:])
+	if err != nil {
+		return err
+	}
+	toDelete = append(toDelete, files...)
+
+	return i.deleteFilesFn(toDelete)
+}
+
+func (i *nsIndex) getCorruptedVolumesForDeletion(filesets []fs.ReadIndexInfoFileResult) ([]string, error) {
+	if len(filesets) <= 1 {
+		return nil, nil
+	}
+
+	// Check for invariants.
+	for j := 0; j+1 < len(filesets); j++ {
+		if !filesets[j].ID.BlockStart.Equal(filesets[j+1].ID.BlockStart) {
+			errorMessage := "all the filesets passed to this function should have the same block start"
+			instrument.EmitAndLogInvariantViolation(i.opts.InstrumentOptions(), func(l *zap.Logger) {
+				l.Error(errorMessage)
+			})
+			return nil, instrument.InvariantErrorf(errorMessage)
+		} else if filesets[j].ID.VolumeIndex >= filesets[j+1].ID.VolumeIndex {
+			errorMessage := "filesets should be ordered by volume index in increasing order"
+			instrument.EmitAndLogInvariantViolation(i.opts.InstrumentOptions(), func(l *zap.Logger) {
+				l.Error(errorMessage)
+			})
+			return nil, instrument.InvariantErrorf(errorMessage)
+		}
+	}
+
+	toDelete := make([]string, 0)
+	hasMoreRecentVolumeOfType := make(map[idxpersist.IndexVolumeType]struct{})
+	// Iterate filesets in reverse order to process higher volume indexes first.
+	for j := len(filesets) - 1; j >= 0; j-- {
+		f := filesets[j]
+
+		// NB: If the fileset info fields contains conflicting information, it means that info file
+		// is missing or corrupted. Delete such filesets, except when it is the most recent volume
+		// in the block.
+		//
+		// The most recent volume is excluded because it is more likely to be actively written to.
+		// If info file writes are not atomic, due to timing readers might observe the file
+		// to be corrupted, even though at that moment the file is being written/re-written.
+		if f.Corrupted && !f.ID.BlockStart.Equal(xtime.FromNanoseconds(f.Info.BlockStart)) {
```

Comment:
What is the meaning behind `f.ID.BlockStart.Equal(xtime.FromNanoseconds(f.Info.BlockStart))` check?

Reply:
Checks for consistency of info file. IIUC, `f.ID.BlockStart` is extracted from filename while `f.Info.BlockStart` is read from info file.